File: src/coxfun.f
C ***
C
      subroutine init_lik(n, loglik, dloglik, d2loglik)

      implicit none

      integer n
      double precision loglik, dloglik(n), d2loglik(n, n)

      loglik = 0.d0
      call dcopy(n, 0.d0, 0, dloglik, 1)
      call dcopy(n * n, 0.d0, 0, d2loglik, 1)

      return
      end

C ***
C
      subroutine obs_lik(what, ns, antrs, totrs, antevents, size,
     &     totevent, eventset, nn, score, 
     &     antcov, covar,
     &     loglik, dloglik)

C +++
C     This subroutine calculates the 'observed' part of the
C     log likelihood function, and its first order
C     partial derivatives.
C +++

      implicit none

      integer what, ns, totrs, totevent, nn, antcov
      integer antrs(ns), antevents(totrs), size(totrs)
      integer eventset(totevent)
      double precision score(nn), covar(nn, antcov)
      double precision loglik, dloglik(antcov)

      integer rsindx, indx, rs, j, i, s, who

      indx = 0
      rsindx = 0
      do rs = 1, ns
         do j = 1, antrs(rs)
            rsindx = rsindx + 1
C +++ The following is not correct for variance estimates!!
C     So we change; "atomic" risksets can be renoverd.
C            if (antevents(rsindx) < size(rsindx)) then
            if (size(rsindx) .ge. 2) then
               do i = 1, antevents(rsindx)
                  indx = indx + 1
                  who = eventset(indx)
                  loglik = loglik + score(who)
                  if (what .ge. 1) then
                     call daxpy(antcov, 1.d0, covar(who, 1), nn,
     &                    dloglik, 1)
                  endif
               enddo
            else
               indx = indx + antevents(rsindx)
            endif
C +++ Next risk set!
         enddo
      enddo
      
      return
      end

C ***
C
      subroutine exp_lik(what, ns, antrs, totrs, antevents, size, 
     &     totsize, riskset, nn, score, 
     &     antcov, covar,
     &     sumdscore, sumd2score,
     &     loglik, dloglik, d2loglik)

C +++
C     This subroutine calculates the 'expected' part of the
C     log likelihood function, and its first and second order
C     partial derivatives.
C ***

      implicit none

      integer what, ns, totrs, totsize, nn, antcov
      integer antrs(ns), antevents(totrs), size(totrs), riskset(totsize)
      double precision score(nn), covar(nn, antcov)

      double precision sumdscore(antcov), sumd2score(antcov, antcov)
      double precision loglik, dloglik(antcov)
      double precision d2loglik(antcov, antcov)

      integer rsindx, indx, rs, j, i, s, m, who
      double precision sumscore

      character*1 UPLO
      parameter (UPLO = 'L')

      rsindx = 0
      indx = 0
      do rs = 1, ns
         do j = 1, antrs(rs)
            rsindx = rsindx + 1
C     Reset to zero:
            sumscore = 0.d0
            if (what .ge. 1) call dcopy(antcov, 0.d0, 0, sumdscore, 1)
            if (what .eq. 2) call dcopy(antcov * antcov, 0.d0, 0, 
     &           sumd2score, 1)
C     Go thru riskset(rs, j):
C +++ Not Removed! We don't go thru 'single' risksets too
C            if (antevents(rsindx) < size(rsindx)) then
            if (size(rsindx) .ge. 2) then
               do i = 1, size(rsindx)
                  indx = indx +  1
                  who = riskset(indx)
                  sumscore = sumscore + score(who)
                  if (what .ge. 1) call daxpy(antcov, score(who), 
     &                 covar(who, 1), nn, sumdscore, 1)
                  if (what .eq. 2) then
                     call dsyr(UPLO, antcov, score(who), 
     &                    covar(who, 1), nn, sumd2score, antcov)
C                     call dger(antcov, antcov, score(who), 
C     &                    covar(who, 1), nn, covar(who, 1), nn, 
C     &                    sumd2score, antcov)
                  endif
               enddo
C     Add into loglik:
               loglik = loglik - antevents(rsindx) * log(sumscore)
               if (what .ge. 1) then
C     Add into dloglik:
                  call daxpy(antcov, -antevents(rsindx) / sumscore,
     &                 sumdscore, 1, dloglik, 1)
               endif
               if (what .eq. 2) then
                  call daxpy(antcov * antcov, 
     &                 antevents(rsindx) / sumscore, sumd2score, 1,
     &                 d2loglik, 1)
                  call dsyr(UPLO, antcov, 
     &                 -antevents(rsindx) / sumscore**2, sumdscore, 1,
     &                 d2loglik, antcov)
 
C                  call dger(antcov, antcov, 
C     &                 -antevents(rsindx) / sumscore**2, sumdscore, 1,
C     &                 sumdscore, 1, d2loglik, antcov)
               endif
            else
               indx = indx + size(rsindx)
            endif
C     Done adding in; next riskset.
         enddo
      enddo
      
      return
      end

C     ***
C     
      subroutine efron_lik(what, ns, antrs, totrs, antevents, size, 
     &     totevent, totsize, eventset, riskset, nn, score, 
     &     antcov, covar,
     &     sumdscore, sumd2score,
     &     loglik, dloglik, d2loglik)

C     +++
C     This subroutine calculates the 'expected' part of the
C     log likelihood function, and its first and second order
C     partial derivatives.
C     ***

      implicit none

      integer what, ns, totrs, totevent, totsize, nn, antcov
      integer antrs(ns), antevents(totrs), size(totrs)
      integer eventset(totevent), riskset(totsize)
      double precision score(nn), covar(nn, antcov)

      double precision sumdscore(antcov), sumd2score(antcov, antcov)
      double precision loglik, dloglik(antcov)
      double precision d2loglik(antcov, antcov)

      integer rsindx, indx, eindx, rs, j, i, r, s, m, who
      double precision sumscore

C     +++
C     Local (note the deviation from strict standard here!):

      double precision escore, edscore(antcov), ed2score(antcov, antcov)
      double precision w, ws

      character*1 UPLO
      parameter (UPLO = 'L')

      double precision temp(antcov)

      rsindx = 0
      indx = 0
      eindx = 0

      do 1000, rs = 1, ns
         do 900, j = 1, antrs(rs)
            rsindx = rsindx + 1
C     Reset to zero:
            sumscore = 0.d0
            escore = 0.d0
            if (what .ge. 1) then
               call dcopy(antcov, 0.d0, 0, sumdscore, 1)
               call dcopy(antcov, 0.d0, 0, edscore, 1)
               if (what .eq. 2) then
                  call dcopy(antcov * antcov, 0.d0, 0, sumd2score, 1)
                  call dcopy(antcov * antcov, 0.d0, 0, ed2score, 1)
               endif
            endif
C     Go thru eventset(rs, j):
C            if (antevents(rsindx) < size(rsindx)) then
            if (size(rsindx) .ge. 2) then
               do 600, i = 1, antevents(rsindx)
                  eindx = eindx +  1
                  who = eventset(eindx)
                  escore = escore + score(who)
                  if (what .ge. 1) then
                     call daxpy(antcov, score(who), covar(who, 1), nn,
     &                    edscore, 1)
                     if (what .eq. 2) then
C                     call dsyr(UPLO, antcov, score(who), 
C     &                    covar(who, 1), 1, ed2score, antcov)

                        call dger(antcov, antcov, score(who), 
     &                       covar(who, 1), nn, covar(who, 1), nn,
     &                       ed2score, antcov)
                     endif
                  endif
 600           continue
C     Go thru riskset(rs, j):
               do 700, i = 1, size(rsindx)
                  indx = indx +  1
C     +++ Removed! We go thru 'single' risksets too
C     if (antevents(rsindx) < size(rsindx)) then
C     Added: skip "atomic" risksets:
                  if (size(rsindx) .ge. 2) then
                     who = riskset(indx)
                     sumscore = sumscore + score(who)
                     if (what .ge. 1) then
                        call daxpy(antcov, score(who), covar(who, 1),
     &                       nn, sumdscore, 1)
                        if (what .eq. 2) then
C                           call dsyr(UPLO, antcov, score(who),
C     &                          covar(who, 1), nn, sumd2score, antcov)
                           call dger(antcov, antcov, score(who),
     &                          covar(who, 1), nn, covar(who, 1), nn,
     &                          sumd2score, antcov)
                        endif
                     endif
                  endif
 700           continue
               
               if (antevents(rsindx) .eq. 1) then
C     Add into loglik:
                  loglik = loglik - log(sumscore)
                  if (what .ge. 1) then
C     Add into dloglik:
                     call daxpy(antcov, -1.d0 / sumscore, 
     &                    sumdscore, 1, dloglik, 1)
                     if (what .eq. 2) then
C     Add into d2loglik:
                        call daxpy(antcov * antcov, 
     &                       antevents(rsindx) / sumscore, 
     &                       sumd2score, 1, d2loglik, 1)
C                        call dsyr(UPLO, antcov, 
C     &                       -antevents(rsindx) / sumscore**2,
C     &                       sumdscore, 1, d2loglik, antcov)
                        call dger(antcov, antcov, 
     &                       -antevents(rsindx) / sumscore**2, 
     &                       sumdscore, 1, sumdscore, 1, 
     &                       d2loglik, antcov)
                        endif
                  endif
C     Done adding in; next riskset.
               else

C     +++ IF TIES:

                  do 800, r = 1, antevents(rsindx)
                     w = dble(r - 1) / dble(antevents(rsindx))
                     ws = w * escore
C     Add into loglik:
                     loglik = loglik - log(sumscore - ws)
                     if (what .ge. 1) then
C     Add into dloglik:
                        call dcopy(antcov, sumdscore, 1, temp, 1)
                        call daxpy(antcov, -w, edscore, 1, temp, 1)
                        call dscal(antcov, 1.d0 / (sumscore -ws), 
     &                       temp, 1)
                        call daxpy(antcov, -1.d0,
     &                       temp, 1, dloglik, 1)
                        if (what .eq. 2) then
C     Add into d2loglik:
C +++ New, next line / Remove again!!:
C                           call dscal(antcov, 1.d0 / (sumscore -ws), 
C     &                          temp, 1)
                           call daxpy(antcov * antcov, 
     &                          1.d0/(sumscore - ws), sumd2score, 1,
     &                          d2loglik, 1)
                           call daxpy(antcov * antcov, 
     &                          -w/(sumscore - ws), ed2score, 1,
     &                          d2loglik, 1)
                           call dsyr(UPLO, antcov, -1.d0, temp, 1,
     &                          d2loglik, antcov)
                        endif
                     endif
 800              continue
C     Done adding in; next riskset.
               endif
            else
               eindx = eindx + antevents(rsindx)
               indx = indx + size(rsindx)
            endif
 900     continue
 1000 continue

      return
      end

C ***
C
      subroutine coxfun(what, method,
     &     totevent, totrs, ns, 
     &     antrs, antevents, size,
     &     totsize, eventset, riskset, 
     &     nn, antcov, covar, offset,
     &     beta,
     &     loglik, dloglik, d2loglik,
     &     score, sumdscore, sumd2score)

C +++ 
C     what     : 0 = Only loglihood.
C                1 = Loglihood and first derivatives.
C                2 = Loglihood, first derivatives, and the negative hessian.
C            other = Nothing.
C     method   : 0 = breslow,
C                1 = efron.
C     totevent : Total number of events.
C     totrs    : Total number of risk sets.
C     ns       : Number of strata.
C
C     antrs     : antrs(i) = No. of risk sets in stratum i, i = 1, ns.
C     antevents : number of events in each riskset.
C     size      : Size of each risk set.
C
C     totsize  : Sum of the risk set sizes.
C     eventset : pointers to events of risk sets (length totevents).
C     riskset  : pointers to members of risk sets (length totsize).
C
c     nn     : No. of spells.
C     antcov : No. of covariates.
C     covar  : matrix of covariates (nn x antcov).
C     offset : Vector of offsets (nn).
C     beta   : Vector of coefficients (antciv).
C
C     loglik   : return value.
C     dloglik  : return value.
C     d2loglik : return value.
C
C     score, sumdscore, sumd2score: 'Work areas', avoiding local
C                                   dynamic memory allocation.
C +++

      implicit none

      integer what, method
      integer totevent, totrs, ns, totsize, nn, antcov
      integer antrs(ns), antevents(totrs), size(totrs)
      integer eventset(totevent), riskset(totsize)
      double precision covar(nn, antcov)
      double precision  offset(nn)

      double precision beta(antcov)

      double precision loglik, dloglik(antcov)
      double precision d2loglik(antcov, antcov)

C +++ Work areas:
      double precision score(nn)
      double precision sumdscore(antcov) 
      double precision sumd2score(antcov, antcov)
C ************************************************************
C     Local:
C
      double precision zero, one
      parameter (zero = 0.d0, one = 1.d0) 
      integer ione
      parameter (ione = 1)
      character*1 trans
      parameter (trans = 'N')

      integer i, s, m

C      integer indx, rsindx

C      double precision sumscore 
C
C *************************************************************

      call init_lik(antcov, loglik, dloglik, d2loglik)

      if ( (what .lt. 0) .or. (what .gt. 2) ) return

C +++ Calculate score(i), i = 1, nn:
      call dcopy(nn, offset, ione, score, ione)
      call dgemv(trans, nn, antcov, one, covar, nn, beta, ione, one,  
     &     score, ione)
      
C +++ 
C     First,add in the 'observed' parts of loglik and dloglik
C     No calculations for d2loglik here (is analytically zero).
C

      call obs_lik(what, ns, antrs, totrs, antevents, size, 
     &     totevent, eventset, nn, score, 
     &     antcov, covar,
     &     loglik, dloglik)

C +++
C     Then, exponentiate 'score':
C
      do i = 1, nn
         score(i) = exp(score(i))
      enddo

C +++
C     and add in the 'expected' parts of loglik, dloglik, and
C     d2loglik.
C
      if (method .eq. 0) then
         call exp_lik(what, ns, antrs, totrs, antevents, size, 
     &        totsize, riskset, nn, score, 
     &        antcov, covar,
     &        sumdscore, sumd2score,
     &        loglik, dloglik, d2loglik)
      else
         call efron_lik(what, ns, antrs, totrs, antevents, size, 
     &        totevent, totsize, eventset, riskset, nn, score, 
     &        antcov, covar,
     &        sumdscore, sumd2score,
     &        loglik, dloglik, d2loglik)
      endif
C +++
C     Fill in the upper triangle part of d2loglik (symmetry!):
C     already done above! (Maybe shouldnt...) Change Back:

      if (what .eq. 2) then

         do s = 1, antcov
            do m = s + 1, antcov
               d2loglik(s, m) = d2loglik(m, s)
            enddo
         enddo

      endif

      
      return
      end

File: src/coxfun.c
#include <R.h>
#include <R_ext/BLAS.h>
#include <Rmath.h>
#include "sup.h"
#include "eha_zeroin.h"
#include "coxfun.h"

extern RS_fun *eha_rs;

static double gam1_fun(double gam, void *info){

    RiskSet *risk;
    double dg, s, egam;
    int i, who;

    risk = info;

    dg = 0.0;
    for (i = 0; i < risk->size; i++){
	who = risk->riskset[i];
	s = score[who]; /* 'score' is global. */
	dg += s;
    }

    egam = exp(gam);
    for (i = 0; i < risk->antevents; i++){
	who = risk->riskset[i];
	s = score[who];
	dg += s / expm1(-s * egam);
    }


    return(-dg);
}

/* static double get1_gam(RiskSet *risk){ An error?! */
void get1_gam(RiskSet *risk){

    /* Binomial, cloglog link */
    int i, itmax;

    double gam, eps;

    double gmin, gmax;
    double ax, bx;

    int who;
    double what;


    if (risk->size == risk->antevents) {
	if (!(risk->out))
	    warning("[get1_gam] gamma positive infinite");
	risk->gamma = 1000.0;
	risk->hazard = 1.0;
	return;
    }else if (risk->size == 1){
	if (!(risk->out))
	    warning("[get1_gam] gamma negative infinite");
	risk->gamma = -1000;
	risk->hazard = 0.0;
	return;
    }

    itmax = 25;
    eps = 0.000001;

    risk->tot_score = 0.0;
    who = risk->riskset[0];
    risk->tot_score += score[who];
    gmin = lin[who];
    gmax = gmin;
    for (i = 1; i < risk->size; i++){
	who = risk->riskset[i];
	risk->tot_score += score[who];
	what = lin[who];
	if (what < gmin){ 
	    gmin = what;
	}else{
	    if (what > gmax) gmax = what;
	}
    }
    if (risk->antevents == 1){
	who = risk->eventset[0];
	/*
	  gam = 1.0 - 
	    R_pow_di(1.0 - score[who] / risk->tot_score, 1.0 / score[who]);
	*/
	gam = log(-log1p(-score[who] / risk->tot_score) / score[who]);
    }else{
	gam = 1.0 - (double)(risk->antevents) / risk->tot_score;
	gam = log(-log1p(-(double)(risk->antevents) / 
			 (double)(risk->size)) ); /* start value */
	ax = gam - gmax;
	bx = gam - gmin;
	if (abs (ax - bx) < eps){
	    risk->gamma = (ax + bx) / 2.0;
	}else{
	    if (gam1_fun(ax, risk) * gam1_fun(bx, risk) > 0.0){
		Rprintf("f(%f) = %f, f(%f) = %f\n", 
			ax, gam1_fun(ax, risk), bx, gam1_fun(bx, risk));
		Rprintf("antevents = %f\n", risk->antevents); 
		Rprintf("size = %f\n", risk->size); 
		error("\nWrong interval in [get0_gam]");
	    }
	    gam = eha_zeroin(ax, bx, &gam1_fun, risk, &eps, &itmax);
	}
    }
    risk->gamma = gam;
    risk->hazard = 1.0 - exp(-exp(gam));
}

void ml_rs(int what, RiskSet *risk,
	   double *b, double e_frac,
	   double *loglik, double *dloglik, double *d2loglik){

    double h1, h11;
    double *h21;

    int i, who;
    double egam, hil, ehil, bil, gil;

    double zero = 0.0;
    double one = 1.0;

    int izero = 0;
    int ione = 1;

    char up = 'U';

    if (risk->out) return;    if (risk->antevents == risk->size) return;

    /* get "gamma[rs]"; in risks[rs]: */
    get1_gam(risk);

/*  Initialize:  */
    h21 = sumdscore;
    h1 = zero;
    h11 = zero;
    F77_CALL(dcopy)(&p, &zero, &izero, h21, &ione);

    egam = exp(risk->gamma);

/* Events : */

    for (i = 0; i < risk->antevents; i++){
	who = risk->eventset[i];
	hil = egam * score[who];
	ehil = exp(-hil);
	/* loglik += log(one - ehil) + hil; */
	*loglik += log1p(-ehil) + hil;

	if (what >= 1){  /* First derivatives */
	    bil = hil / (one - ehil);
	    /* bil = -hil / expm1(-hil); */
            h1 += bil;
            /* Update h2[j] "+= x[j, who] * bil": */
	    F77_CALL(daxpy)(&p, &bil, (x + p * who), &ione, 
			    dloglik, &ione);
	    
	    if (what >= 2){ /* Second derivatives */
		gil = bil * (ehil + hil * ehil - one) / (one - ehil); 
		    /* gil = bil * (bil * ehil - one); */
		h11 += gil;
		/* Update h21: */
		F77_CALL(daxpy)(&p, &gil, (x + p * who), &ione, 
				h21, &ione);
		/* Update h22 (upper triangle): */
		F77_CALL(dsyr)(&up, &p, &gil, (x + p * who), &ione, 
			       d2loglik, &p);
	    }
	}
    }

/* All in riskset: */
    for (i = 0; i < risk->size; i++){
	who = risk->riskset[i];
	hil = -egam * score[who]; /* Note 1! */
	ehil = exp(hil);
	*loglik += hil;
	if (what >= 1){ /* First derivatives */
	    h1 += hil;
	    F77_CALL(daxpy)(&p, &hil, (x + p * who), &ione,
			    dloglik, &ione);

	    if (what >= 2){ /* Second derivatives: */
		hil = -hil; /* Note 2! */
		h11 += hil; /* Added 0.99-15 */
		F77_CALL(daxpy)(&p, &hil, (x + p * who), &ione, 
				h21, &ione);
		F77_CALL(dsyr)(&up, &p, &hil, (x + p * who), &ione,
			       d2loglik, &p);
	    }
	}
    }

    if (what >= 2){
	h11 = -one / h11;
	F77_CALL(dsyr)(&up, &p, &h11, h21, &ione, d2loglik, &p);
    }
}

static void cox_obs_rs(int what, RiskSet *risk,
		       double *b,
		       double *loglik, double *dloglik){

    int i, who;
    double one = 1.0;
    int ione = 1;

    if (risk->out) return;

    for (i = 0; i < risk->antevents; i++){
	who = risk->eventset[i];
	*loglik += lin[who];
	if (what >= 1) F77_CALL(daxpy)(&p, &one, (x + p * who), &ione,
				       dloglik, &ione);
    }
}
    
void breslow_rs(int what, RiskSet *risk, 
		double *b, double e_frac,
		double *loglik, double *dloglik, 
		double *d2loglik){

/* Calculates the "breslow" variation of the partial likelihood, */
/* and eventually its first and second partial derivatives.      */

    int i, who;
    char up = 'U';

    double zero = 0.0;
    int izero = 0;
    int ione = 1;
    double sumscore;
    int p2 = p * p;
    double alpha;
    
    if (risk->out) return;

    /* First the "observed" part (common to the 'efron' method): */
    
    cox_obs_rs(what, risk,
	       b,
	       loglik, dloglik);
    
    /* Then the "expected" part: */

    /* Initialize: */

    sumscore = 0.0;
    if (what >= 1){
	F77_CALL(dcopy)(&p, &zero, &izero, sumdscore, &ione);
	if (what >= 2){
	    F77_CALL(dcopy)(&p2, &zero, &izero, sumd2score, &ione);
	}
    }

    /* Go thru riskset: */

    for (i = 0; i < risk->size; i++){
	who = risk->riskset[i];
	sumscore += score[who];
	if (what >= 1){ /* First derivatives: */
	    F77_CALL(daxpy)(&p, (score + who), (x + p * who), &ione, 
			    sumdscore, &ione);

	    if (what >= 2){ /* Second derivatives: */
		F77_CALL(dsyr)(&up, &p, (score + who), 
			       (x + p * who), &ione,
			       sumd2score, &p);
	    }
	}
    }

    /* Add in: */

    *loglik -= risk->antevents * log(sumscore);
    if (what >= 1){
	alpha = -(double)(risk->antevents) / sumscore;
	F77_CALL(daxpy)(&p, &alpha, sumdscore, &ione, 
			dloglik, &ione);
	if (what >= 2){
	    alpha = -alpha;
	    F77_CALL(daxpy)(&p2, &alpha, sumd2score, &ione,
			    d2loglik, &ione);
	    alpha = -alpha / sumscore;
	    F77_CALL(dsyr)(&up, &p, &alpha, sumdscore, &ione,
			   d2loglik, &p);
	}
    }

}

void efron_rs(int what, RiskSet *risk, 
	      double *b, double e_frac,
			 double *loglik, double *dloglik, 
			 double *d2loglik){
  
    /*************************************************************
C     This subroutine calculates the 'efron' variation of the
C     log likelihood function, and its first and second order
C     partial derivatives.
    *************************************************************/

    int i, r, who;
    double sumscore;

/*************************************************************
C     +++
C     Local (note the deviation from strict standard here!):
**************************************************************/

    double escore;
    double *edscore, *ed2score;
    double w, ws;

    char up = 'U';

    double *temp; /*(antcov)*/
    int p2 = p * p;

    int izero = 0;
    int ione = 1;

    double zero = 0.0;
    double one = 1.0;

    double alpha;

    if (risk->out) return;

    /* First the "observed" part (common to the 'breslow' method): */

    cox_obs_rs(what, risk,
	       b,
	       loglik, dloglik);

    /* Then the "expected" part: */

    edscore = Calloc(p, double);
    ed2score = Calloc(p2, double);

    temp = Calloc(p, double);
		     

/*     Reset to zero: */
    sumscore = zero;
    escore = zero;
    if (what >= 1){
	F77_CALL(dcopy)(&p, &zero, &izero, sumdscore, &ione);
        F77_CALL(dcopy)(&p, &zero, &izero, edscore, &ione);
	if (what >= 2){
	    F77_CALL(dcopy)(&p2, &zero, &izero, sumd2score, &ione);
	    F77_CALL(dcopy)(&p2, &zero, &izero, ed2score, &ione);
	}
    }

/*     Go thru riskset(rs, j): */
    for (i = 0; i < risk->size; i++){
	who = risk->riskset[i];
	sumscore += score[who];
	if (what >= 1){
	    F77_CALL(daxpy)(&p, (score + who), (x + p * who),
			    &ione, sumdscore, &ione);
	    if (what >= 2){
		F77_CALL(dsyr)(&up, &p, (score + who),
                               (x + p * who), &ione, 
                               sumd2score, &p);
/*		F77_CALL(dger)(&p, &p, (score + who), (x + p * who), &ione,
		(x + p * who), &ione, sumd2score, &p); */
		
	    }
	}
    }
               
    if (risk->antevents == 1){ /* No ties */
/*     Add into loglik: */
	*loglik -= log(sumscore);
	if (what >= 1){
/*     Add into dloglik: */
	    alpha = -one / sumscore;
	    F77_CALL(daxpy)(&p, &alpha, 
			    sumdscore, &ione, dloglik, &ione);
	    if (what >= 2){
/*     Add into d2loglik: */
		alpha = (double)risk->antevents / sumscore; 
		F77_CALL(daxpy)(&p2, &alpha, sumd2score, &ione, 
				d2loglik, &ione);
		alpha = -(double)(risk->antevents) / 
		    (sumscore * sumscore);
		F77_CALL(dsyr)(&up, &p, &alpha, sumdscore, &ione, 
		d2loglik, &p); 
/*		F77_CALL(dger)(&p, &p, &alpha, sumdscore, &ione, 
			       sumdscore, &ione, d2loglik, &p); */
	    }
	}
    }else{
/*     +++ IF TIES: */
	
/* Go thru events and create escore, edscore and ed2score: */
	for (i = 0; i < risk->antevents; i++){
	    who = risk->eventset[i];
	    escore = escore + score[who];
	    if (what >= 1){ /* first derivatives */
		F77_CALL(daxpy)(&p, (score + who), (x + p * who), &ione,
				edscore, &ione);
		if (what >= 2){ /* second derivatives */
		    F77_CALL(dsyr)(&up, &p, (score + who), 
				   (x + p * who), &ione,
				   ed2score, &p); 
/*	       F77_CALL(dger)(&p, &p, (score + who), (x + p * who), &ione,
	       (x + p * who), &ione, ed2score, &p); */
		}
	    }
	}


	for (r = 0; r < risk->antevents; r++){
/* WRONG: (Fortran)   w = (double)(r - ione) / (double)(risk->antevents);*/
	    w = (double)r / (double)(risk->antevents);
	    ws = w * escore;
/*     Add into loglik: */
	    *loglik -= log(sumscore - ws);
	    if (what >= 1){
/*     Add into dloglik: */
		F77_CALL(dcopy)(&p, sumdscore, &ione, temp, &ione);
		alpha = -w;
		F77_CALL(daxpy)(&p, &alpha, edscore, &ione, temp, &ione);
		alpha = one / (sumscore - ws);
		F77_CALL(dscal)(&p, &alpha, temp, &ione);
		alpha = -one;
		F77_CALL(daxpy)(&p, &alpha, temp, &ione, dloglik, &ione);
		if (what >= 2){
/*     Add into d2loglik: */
		    alpha = one / (sumscore - ws);
		    F77_CALL(daxpy)(&p2, &alpha, sumd2score, &ione,
				    d2loglik, &ione);
		    alpha = -w / (sumscore - ws);
		    F77_CALL(daxpy)(&p2, &alpha, ed2score, &ione,
				    d2loglik, &ione);
		    alpha = -one;
		    F77_CALL(dsyr)(&up, &p, &alpha, temp, &ione,
				   d2loglik, &p);
		}
	    }
	}
    }
    Free(temp);
    Free(ed2score);
    Free(ed2score);
}


void mppl_rs(int what, RiskSet *risk,
	     double *b, double e_frac,
	     double *loglik, double *dloglik, double *d2loglik){

    if (risk->antevents == risk->size) return;
    if (risk->out) return;
    if (risk->antevents == 1){
	breslow_rs(what, risk,
		   b, e_frac,
		       loglik, dloglik, d2loglik);
    }else if (risk->antevents <= e_frac * risk->size){
	efron_rs(what, risk, 
		 b, e_frac,
		 loglik, dloglik, d2loglik);
    }else{
	ml_rs(what, risk,
	      b, e_frac,
	      loglik, dloglik, d2loglik);
    }
}

void coxfun(int what, int totrs, RiskSet *risks, 
/*	    int ml, int method, */ double e_frac, 
	    double *b, 
	    double *loglik, double *dloglik, double *d2loglik){

    /* Note that the following is deprecated:      */
    /* ml : 0 = "Cox", 1 = "Maximum likelihood"    */
    /* method: 0 = "efron",  if ml = 0,            */ 
    /*         1 = "breslow, if ml = 0             */
    /*                                             */
    /* method: 0 = "mppl",   if ml = 1,            */ 
    /*         1 = "ml,      if ml = 1             */
    /*                                             */


    int izero = 0;
    int ione = 1;
    double zero = 0.0;
    double one = 1.0;
    char trans = 'T';

    int p2 = p * p;
    int s, m, rs;

    /* Initialize: */

    *loglik = 0.0;
    if (what < 0) return;

    if (what >= 1){
	F77_CALL(dcopy)(&p, &zero, &izero, dloglik, &ione);
	if (what >= 2){
	    F77_CALL(dcopy)(&p2, &zero, &izero, d2loglik, &ione);  
	}
    }

    /* Calculate 'lin' and 'score' */

    F77_CALL(dcopy)(&nn, offset, &ione, lin, &ione);
    F77_CALL(dgemv)(&trans, &p, &nn, &one, x, 
		    &p, b, &ione, &one, lin, &ione); 
    for (m = 0; m < nn; m++) score[m] = exp(lin[m]);

    /* Start walking thru risksets: */
    
    for (rs = 0; rs < totrs; rs++){
	eha_rs(what, (risks + rs),
	       b, e_frac, 
	       loglik, dloglik, d2loglik);
    }
	/* Fill in the lower part of 'd2loglik': */

    if (what >= 2){
	for (s = 0; s < p; s++){
	    for (m = 0; m < s; m++){
		d2loglik[s + m * p] = d2loglik[m + s * p];
	    }
	}
    }
}

